ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java

```diff
@@ -13,119 +13,538 @@
+   * @throws Exception Any unexpected exception.
    */
   @Test
   public void badIdTest() throws Exception {
-    MockClusterMap clusterMap = new MockClusterMap();
-    List<String> blobIdLikes = new ArrayList<>();
+    for (Short version : versions) {
+      generateAndAssertBadBlobId(version);
+    }
+  }
+
+  /**
+   * Tests blobId long form in both v1 and v2.
+   */
+  @Test
+  public void testBlobIdLongForm() {
+    assertBlobIdLongForm(BLOB_ID_V1);
+    assertBlobIdLongForm(BLOB_ID_V2);
+  }
+
+  /**
+   * Makes a blobId-like string in the same way as used in blobId serialization, and then deserialize from the
+   * string to construct blobId object.
+   * @param version The version of BlobId.
+   * @throws Exception Any unexpected exception.
+   */
+  private void composedBlobIdAndDeserialize(short version) throws Exception {
+    String srcUUID = UUID.randomUUID().toString();
+    String srcBlobIdStr;
+    switch (version) {
+      case BLOB_ID_V1:
+        srcBlobIdStr =
+            buildBlobIdLike(version, null, null, null, null, referencePartitionId, srcUUID.length(), srcUUID);
+        break;
+
+      case BLOB_ID_V2:
+        srcBlobIdStr =
+            buildBlobIdLike(version, referenceFlag, referenceDatacenterId, referenceAccountId, referenceContainerId,
+                referencePartitionId, srcUUID.length(), srcUUID);
+        break;
+
+      default:
+        throw new IllegalArgumentException("Invalid version number blob" + version);
+    }
+    deserializeBlobIdAndAssert(version, srcBlobIdStr, referencePartitionId);
+  }
 
-    short goodVersion = 1;
-    PartitionId goodPartitionId = clusterMap.getWritablePartitionIds().get(0);
+  /**
+   * Construct a blobId, serialize it into string, and then deserialize from the string to construct BlobId object.
+   * @param version The version of BlobId.
+   * @throws Exception Any unexpected exception.
+   */
+  private void serDesBlobId(short version) throws Exception {
+    String srcBlobIdStr;
+    switch (version) {
+      case BLOB_ID_V1:
+        srcBlobIdStr = new BlobIdBuilder(referencePartitionId).build().getID();
+        break;
+
+      case BLOB_ID_V2:
+        srcBlobIdStr = new BlobIdBuilderV2(referencePartitionId).setFlag(referenceFlag)
+            .setDatacenterId(referenceDatacenterId)
+            .setAccountId(referenceAccountId)
+            .setContainerId(referenceContainerId)
+            .build()
+            .getID();
+        break;
+
+      default:
+        throw new IllegalArgumentException("invalid version number blob" + version);
+    }
+    deserializeBlobIdAndAssert(version, srcBlobIdStr, referencePartitionId);
+  }
+
+  /**
+   * Generates bad blobId strings, and deserializes from the string.
+   * @param version The version of BlobId.
+   * @throws Exception Any unexpected exception.
+   */
+  private void generateAndAssertBadBlobId(Short version) throws Exception {
+    List<String> blobIdLikes = new ArrayList<>();
```

Comment:
Changed to `invalidBlobIdLikeList`. Prefer to keep `Like` because the ids are hand-made.